src/main/java/com/uw/controller/TrainerController.java

```diff
@@ -0,0 +1,229 @@
+        if (!isAuthenticated) {
+            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("Invalid credentials");
+        }
+
+        if (!authService.isTrainer(providedUsername, providedPassword)) {
+            return ResponseEntity.status(HttpStatus.NOT_FOUND).body("Trainer not found");
+        }
+
+        Trainer trainer = trainerService.findTrainerByUsername(username);
+
+        TrainerDTO trainerProfileDTO = TrainerMapper.toDTO(trainer);
+
+        return ResponseEntity.ok().header(HttpHeaders.CONTENT_TYPE, "application/json").body(trainerProfileDTO);
+
+    }
+
+    @PostMapping
+    public ResponseEntity<String[]> create(@RequestBody TrainerDTO trainerDTO) {
+        // Validar entrada
+        if (trainerDTO.getSpecialization() == null || trainerDTO.getFirstName() == null || trainerDTO.getLastName() == null) {
+            return ResponseEntity.badRequest().body(new String[]{"Error: Missing required fields"});
+        }
+
+        TrainingType specialization = new TrainingType();
+        specialization.setTypeName(trainerDTO.getSpecialization());
+        if (specialization.getTypeName() == null) {
+            return ResponseEntity.badRequest().body(new String[]{"Error: Invalid specialization"});
+        }
+
+        // Crear usuario
+        String username = userNameGenerator.generateUsername(trainerDTO.getFirstName(), trainerDTO.getLastName());
+        String password = passwordGenerator.generatePassword(10);
+        User user = new User();
+        user.setUsername(username);
+        user.setPassword(password);
+        user.setFirstName(trainerDTO.getFirstName());
+        user.setLastName(trainerDTO.getLastName());
+        user.setActive(true);
+        // Crear entrenador
+        Trainer trainer = TrainerMapper.toEntity(trainerDTO, specialization, user);
+        trainer.setUser(user);
+        trainerService.createTrainer(trainer); // Guardar entrenador
+
+        // Responder con nombre de usuario y contraseña
+        return ResponseEntity.ok(new String[]{username, password});
+    }
+
+    @PutMapping("/{username}")
+    public ResponseEntity<?> update(
+            @RequestBody TrainerDTO trainerDTO,
+            @PathVariable("username") String username,
+            @RequestHeader(value = "Authorization") String auth){
+
+        String[] credentials = AuthService.decodeCredentials(auth);
+        if (credentials.length != 2) {
+            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("Invalid Authorization header");
+        }
+
+        String providedUsername = credentials[0];
+        String providedPassword = credentials[1];
+
+        // Autenticar el usuario
+        boolean isAuthenticated = authService.authentication(providedUsername, providedPassword);
+        if (!isAuthenticated) {
+            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("Invalid credentials");
```

Comment:
All constants like "Invalid credentials" should be part of a java class to centralize them